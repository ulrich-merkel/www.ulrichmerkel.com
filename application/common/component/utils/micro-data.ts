/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
