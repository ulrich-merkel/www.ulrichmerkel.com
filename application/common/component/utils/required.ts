/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
