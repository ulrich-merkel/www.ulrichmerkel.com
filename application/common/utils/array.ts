/**
 * Es6 module for handling array related utilities.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */

/**
 * Check if given "input" is a valid array with at least one entry.
 *
 * @param {Array<*>} array - The array to be checked
 * @returns {boolean} Whether this is an array with some entries
 */
export function isValidArray(array?: any[]): boolean {
    return Array.isArray(array) && array.length > 0;
}
