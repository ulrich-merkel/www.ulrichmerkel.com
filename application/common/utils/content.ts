/**
 * Es6 module for helper component.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { get, isArray, isObject, isString } from 'lodash';

import { AVAILABLE_LOCALES } from '../state/intl/duck';

/**
 * Walk through array
 *
 * @private
 * @param {Array} source - The source content
 * @param {Function} callback - The ready callback
 * @returns {Array}
 */
function traverseArray(source, callback) {
    return source.map((value) => {
        return traverse(value, callback); // eslint-disable-line no-use-before-define
    });
}

/**
 * Walk through object
 *
 * @private
 * @param {object} source - The source content
 * @param {Function} callback - The ready callback
 * @returns {object}
 */
function traverseObject(source, callback) {
    const result = {};
    Object.keys(source).forEach((key) => {
        result[key] = traverse(source[key], callback); // eslint-disable-line no-use-before-define
    });
    return result;
}

/**
 * Walk through config
 *
 * @TODO Use functional style, immutable
 *
 * @private
 * @param {object|Array|string} source - The source content
 * @param {Function} callback - The ready callback
 * @returns {*}
 */
function traverse(source, callback) {
    if (isArray(source)) {
        return traverseArray(source, callback);
    }
    if (isObject(source)) {
        return traverseObject(source, callback);
    }
    if (isString(source)) {
        return callback.apply(this, [source]);
    }
    return source;
}

/**
 * @TODO Use functional style, immutable
 *
 * @private
 * @param {object} configContent - The complete content config
 * @param {string} configTranslation - The complete content translation
 * @returns {object}
 */
function translateContent(configContent, configTranslation) {
    if (!configTranslation) {
        return configContent;
    }

    return traverse(configContent, (value) => {
        const translatedString = configTranslation[value];
        if (translatedString) {
            return translatedString;
        }
        return value;
    });
}

/**
 * @param {object} config - The translated config
 * @returns {Function}
 */
function getContentSection(config) {
    return function getContentSectionByKey(key) {
        return get(config, key);
    };
}

/**
 * @TODO use memoize https://addyosmani.com/blog/faster-javascript-memoization/
 *
 * @param {string} [locale=defaultLocale] - The current locale
 * @param {object} config - The translation config
 * @param {string} configKey - The config key
 * @returns {object}
 */
function getTranslatedContent(
    locale = AVAILABLE_LOCALES[0],
    config,
    configKey
) {
    const configContent = get(config, 'content.data');
    const configContentByKey = get(configContent, configKey, configContent);
    const configTranslation = get(config, `${locale.toLowerCase()}.data`);

    return translateContent(configContentByKey, configTranslation);
}

export { getContentSection, getTranslatedContent };
