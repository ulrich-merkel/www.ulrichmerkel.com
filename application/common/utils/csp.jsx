/* eslint-disable immutable/no-let */
/**
 * Es6 module to handle content security policies.
 *
 * Mitigate the risk of cross-site scripting and other content-injection
 * attacks.
 *
 * This can be done by setting a `Content Security Policy` which
 * whitelists trusted sources of content for your website.
 *
 * The example header below allows ONLY scripts that are loaded from
 * the current website's origin (no inline scripts, no CDN, etc).
 * That almost certainly won't work as-is for your website!
 *
 * To make things easier, you can use an online CSP header generator
 * such as: http://cspisawesome.com/.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link https://content-security-policy.com/}
 * @see {@link http://www.w3.org/TR/CSP11/)}
 * @see {@link http://www.html5rocks.com/en/tutorials/security/content-security-policy/}
 * @see {@link https://www.thepolyglotdeveloper.com/2015/03/create-a-random-nonce-string-using-javascript/}
 * @see {@link https://blogs.dropbox.com/tech/2015/09/unsafe-inline-and-nonce-deployment/}
 * @see {@link https://report-uri.io/home/generate}
 *
 * @requires lodash
 *
 * @changelog
 * - 0.0.1 basic function and structure
 */
import { isEmpty } from 'lodash';

/**
 * Generate unique strings to be used as nonce
 * (means number only used once)
 *
 * @private
 * @param {number} length - The generated number length
 * @returns {string}
 */
function getNonce(length = 40) {
    const possibleChars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    const text = [];

    for (let i = 0; i < length; i = i + 1) {
        text.push(possibleChars.charAt(Math.floor(Math.random() * possibleChars.length)));
    }

    return text.join('');
}

/**
 * Generate nonce config for inline styles
 * and scripts.
 *
 * @returns {object}
 */
function getNonceConfig() {
    return {
        style: {
            base: getNonce()
        },
        script: {
            bootstrap: getNonce(),
            config: getNonce()
        }
    };
}

/**
 * Generate CSP header to improve security, relying on nonce for dynamically
 * generated styles and scripts.
 *
 * 'unsafe-inline' is a hack for safari, which don't support nonces.
 * 'data:' is used for inline base64 placeholder images.
 *
 * @see {@link http://stackoverflow.com/questions/32788355/csp-nonce-ignored-by-safari}
 *
 * @param {object} nonceConfig - The output from a getNonceConfig call
 * @returns {string}
 */
function getCspRules(nonceConfig) {
    if (isEmpty(nonceConfig)) {
        return '';
    }

    const rules = [];
    /* eslint-disable quotes */
    rules.push(`default-src 'self' www.ulrichmerkel.com;`);
    rules.push(`script-src 'self' www.ulrichmerkel.com 'nonce-${nonceConfig.script.bootstrap}' 'nonce-${nonceConfig.script.config}' 'unsafe-inline';`);
    rules.push(`style-src 'self' www.ulrichmerkel.com 'nonce-${nonceConfig.style.base}' 'unsafe-inline';`);
    rules.push(`img-src 'self' www.ulrichmerkel.com data:;`);
    rules.push(`child-src 'self';`);
    rules.push(`object-src 'none';`);
    /* eslint-enable quotes */

    return rules.join('');
}

export {
    getNonceConfig,
    getCspRules
};
