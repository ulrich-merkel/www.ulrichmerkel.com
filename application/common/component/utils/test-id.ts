/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
