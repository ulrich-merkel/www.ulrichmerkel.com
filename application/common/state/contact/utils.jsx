/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 *
 * @TODO Should be refactored and moved to actions
 *
 * @requires common/utils/validation
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';

/**
 * Validation helper to check all inputs.
 *
 * @function
 * @param {object} [state={}] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
function validate(state = {}) {
    return {
        name: required(state.name),
        email: required(state.email) && email(state.email),
        subject: required(state.subject),
        message: required(state.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @function
 * @param {object} state - The current redux contact state
 * @returns {boolean} The validation status
 */
function isValid(state) {
    const validated = validate(state);

    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

function canSendForm(state) {
    const isStateValid = isValid(state);

    return isStateValid
        && state.pristine
        && !state.pending
        && isBrowser(state);
}

export {
    canSendForm,
    isValid,
    validate
};
