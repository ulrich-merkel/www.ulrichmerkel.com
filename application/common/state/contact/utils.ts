/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';
import { ContactFormType, ValidateType } from './types';

/**
 * Validation helper to check all inputs.
 *
 * @param {object} [contactForm] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
export function validate(contactForm: ContactFormType): ValidateType {
    return {
        name: required(contactForm?.name),
        email: required(contactForm?.email) && email(contactForm?.email),
        subject: required(contactForm?.subject),
        message: required(contactForm?.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} The validation status
 */
export function isValid(contactForm: ContactFormType): boolean {
    const validated = validate(contactForm);
console.log({validated, contactForm})
    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

/**
 * Check if the form can be send.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} Whether the submit button is enabled or not
 */
export function canSendForm(contactForm: ContactFormType): boolean {
    const isStateValid = isValid(contactForm);

    return (
        isStateValid &&
        contactForm.pristine &&
        !contactForm.pending &&
        isBrowser(contactForm)
    );
}
