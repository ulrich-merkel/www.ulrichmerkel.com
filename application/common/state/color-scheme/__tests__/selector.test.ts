import { mockState } from '../../../../__tests__/utils/get-mocked-store';
import {
    selectStateColorScheme,
    selectStateColorSchemeSelected
} from '../selector';

describe('selectStateColorScheme', function fnDescribe() {
    it('should return the correct state', function fnIt() {
        expect(selectStateColorScheme(mockState)).toMatchSnapshot();
    });
    it('should return the initial state if resource key is not found', function fnIt() {
        const state = {
            foo: {
                bar: 'lorem'
            }
        };
        expect(selectStateColorScheme(state)).toMatchSnapshot();
    });
    it('should return the initial state if no state is found', function fnIt() {
        expect(selectStateColorScheme()).toMatchSnapshot();
    });
});

describe('selectStateColorSchemeSelected', function fnDescribe() {
    it('should return the correct state', function fnIt() {
        expect(selectStateColorSchemeSelected(mockState)).toMatchSnapshot();
    });
    it("should return the default if state isn't found", function fnIt() {
        const state = {
            foo: {
                bar: 'lorem'
            }
        };
        expect(selectStateColorSchemeSelected(state)).toMatchSnapshot();
    });
});
