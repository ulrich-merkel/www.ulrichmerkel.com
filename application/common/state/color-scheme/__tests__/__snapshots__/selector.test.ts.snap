// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectStateColorScheme should return the correct state 1`] = `
Object {
  "meta": Object {
    "isInitial": false,
  },
  "payload": Object {
    "selected": "light",
  },
}
`;

exports[`selectStateColorScheme should return the initial state if no state is found 1`] = `
Object {
  "meta": Object {
    "isInitial": true,
  },
  "payload": Object {
    "selected": undefined,
  },
}
`;

exports[`selectStateColorScheme should return the initial state if resource key is not found 1`] = `
Object {
  "meta": Object {
    "isInitial": true,
  },
  "payload": Object {
    "selected": undefined,
  },
}
`;

exports[`selectStateColorSchemeSelected should return the correct state 1`] = `"light"`;

exports[`selectStateColorSchemeSelected should return the default if state isn't found 1`] = `undefined`;
